homeassistant/components/reolink/entity.py

```diff
@@ -24,11 +28,9 @@ class ReolinkBaseCoordinatorEntity(CoordinatorEntity):
     def __init__(
         self,
         reolink_data: ReolinkData,
-        coordinator: DataUpdateCoordinator | None = None,
+        coordinator: DataUpdateCoordinator[_T],
     ) -> None:
         """Initialize ReolinkBaseCoordinatorEntity for a NVR entity without a channel."""
-        if coordinator is None:
-            coordinator = reolink_data.device_coordinator
```

Comment:
Please leave the default behaviour if no coordinator is specified, this is used in PR https://github.com/home-assistant/core/pull/87943

Reply:
This makes it quite difficult if not impossible to type unfortunately. The `coordinator` argument would need to be `DataUpdateCoordinator[_T] | None` but you can't assign the `device_coordinator` to it as it's `DataUpdateCoordinator[None]`. We would have to either use `cast` or two type ignores here.

For #87943 it would probably be easier to explicitly pass `relink_data.device_coordinator` as argument.

Reply:
Well that highlights a even bigger problem: if PR https://github.com/home-assistant/core/pull/87943 passes in relink_data.device_coordinator it will then also fail the mypy test since the type is DataUpdateCoordinator[None] and not DataUpdateCoordinator[_T] wouldn't it?

Schouldn't the type then be DataUpdateCoordinator[_T] | DataUpdateCoordinator[None] | None?
Or to make our lives a bit easier: DataUpdateCoordinator[Any] | None

Reply:
> Well that highlights a even bigger problem: if PR #87943 passes in relink_data.device_coordinator it will then also fail the mypy test since the type is DataUpdateCoordinator[None] and not DataUpdateCoordinator[_T] wouldn't it?

No, that will work. That's similar to the `super().__init__` call in `ReolinkUpdateEntity` here.
It's only an issue inside `ReolinkBaseCoordinatorEntity.__init__` as there `coordinator` needs to be a generic, however from the outside, we pass a concrete type so that work just fine.

> Schouldn't the type then be DataUpdateCoordinator[_T] | DataUpdateCoordinator[None] | None? Or to make our lives a bit easier: DataUpdateCoordinator[Any] | None

`DataUpdateCoordinator[Any]` would work but that would also disable the type checking for it.

Reply:
I would vote for using DataUpdateCoordinator[Any] or DataUpdateCoordinator[_T] | DataUpdateCoordinator[None] | None.
I think having a default and simplifying the platform code would be better than simplifying the mypy typing.

Reply:
If you want to, we can always subclass `ReolinkBaseCoordinatorEntity` and pass the default that way. similarly to `ReolinkCoordinatorEntity`. That does work.

I just didn't do it here, as `ReolinkBaseCoordinatorEntity` is only instantiated once explicitly with a value for `coordinator`. The other time is via `super().__init__` but we pass an argument there already.